numpy/_core/src/umath/string_ufuncs.cpp

```diff
@@ -1065,26 +928,61 @@ init_string_ufuncs(PyObject *umath)
     dtypes[1] = NPY_DEFAULT_INT;
     if (init_ufunc(
             umath, "str_len", 1, 1, dtypes, ENCODING::ASCII,
-            string_len_loop<ENCODING::ASCII>, NULL, NULL) < 0) {
+            string_str_len_loop<ENCODING::ASCII>, NULL, NULL) < 0) {
         return -1;
     }
     if (init_ufunc(
             umath, "str_len", 1, 1, dtypes, ENCODING::UTF32,
-            string_len_loop<ENCODING::UTF32>, NULL, NULL) < 0) {
+            string_str_len_loop<ENCODING::UTF32>, NULL, NULL) < 0) {
         return -1;
     }
 
     dtypes[0] = NPY_OBJECT;
     dtypes[1] = NPY_BOOL;
-    if (init_ufunc(
-            umath, "isalpha", 1, 1, dtypes, ENCODING::ASCII,
-            string_isalpha_loop<ENCODING::ASCII>, NULL, NULL) < 0) {
-        return -1;
-    }
-    if (init_ufunc(
-            umath, "isalpha", 1, 1, dtypes, ENCODING::UTF32,
-            string_isalpha_loop<ENCODING::UTF32>, NULL, NULL) < 0) {
-        return -1;
+
+    const char *unary_buffer_method_names[] = {
+        "isalpha", "isalnum", "isdigit", "isspace", "islower",
+        "isupper", "istitle", "isdecimal", "isnumeric",
+    };
+
+    static buffer_method<ENCODING::ASCII> unary_buffer_ascii_methods[] = {
+        &Buffer<ENCODING::ASCII>::isalpha,
+        &Buffer<ENCODING::ASCII>::isalnum,
+        &Buffer<ENCODING::ASCII>::isdigit,
+        &Buffer<ENCODING::ASCII>::isspace,
+        &Buffer<ENCODING::ASCII>::islower,
+        &Buffer<ENCODING::ASCII>::isupper,
+        &Buffer<ENCODING::ASCII>::istitle,
+    };
+
+    static buffer_method<ENCODING::UTF32> unary_buffer_utf32_methods[] = {
+        &Buffer<ENCODING::UTF32>::isalpha,
+        &Buffer<ENCODING::UTF32>::isalnum,
+        &Buffer<ENCODING::UTF32>::isdigit,
+        &Buffer<ENCODING::UTF32>::isspace,
+        &Buffer<ENCODING::UTF32>::islower,
+        &Buffer<ENCODING::UTF32>::isupper,
+        &Buffer<ENCODING::UTF32>::istitle,
+        &Buffer<ENCODING::UTF32>::isdecimal,
+        &Buffer<ENCODING::UTF32>::isnumeric,
+    };
+
+    for (int i = 0; i < 9; i++) {
+        if (i < 7) { // isdecimal & isnumeric do not support ASCII
```

Comment:
I don't really mind either way, but is there a reason that `isdecimal` and `isnumeric` are not supported?

Reply:
Because the python3 APIs we're implementing are unicode-aware:

https://docs.python.org/3/library/stdtypes.html#str.isdecimal
https://docs.python.org/3/library/stdtypes.html#str.isnumeric

Thankfully they're not locale-aware, that would be way worse.

Reply:
I guess in principle one could support it right? I'm totally fine not doing so, though!